### PR TITLE
[ROCm] Enable tl.reduce usage on ROCm

### PR DESCRIPTION
**Pull request by jataylo**
Revert aten.prod explicit fallback on ROCm and enabling the use of tl.reduce in triton codegen. This PR also enables an optimisation that was previously conditionalised out for ROCm https://github.com/pytorch/pytorch/pull/102444

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78

### CODE REVIEW

File: torch/_inductor/lowering.py
Comment (peterbell10) [inline torch/_inductor/lowering.py:1735]: Can you also remove everything that references `TRITON_HAS_REDUCE` in `triton_helpers.py`

**Comment by jataylo** [comment]
Unrelated failure [linux-vulkan-bionic-py3.11-clang9 (https://hud.pytorch.org/pr/pytorch/pytorch/104099#14548871639)https://hud.pytorch.org/failure/Makefile%3A145%3A%20recipe%20for%20target%20'all'%20failed

Opening for review, @peterbell10 @ngimel please let me know if there's anything I've missed with permitting tl.reduce and enabling https://github.com/pytorch/pytorch/pull/102444 on ROCm

**Review by peterbell10 — APPROVED** [review]
LGTM

**Comment by jithunnair-amd** [comment]
@malfet Just waiting on your approval before merge

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**
